recipes_source/recipes/profiler.py

```diff
@@ -0,0 +1,203 @@
+"""
+PyTorch Profiler
+====================================
+This recipe explains how to use PyTorch profiler and measure time and
+memory consumption of model's operators.
+
+Introduction
+------------
+PyTorch includes a simple profiler API that is useful when user needs
+to determine the most expensive operators in the model.
+
+In this recipe, we will use a simple Resnet model to demonstrate how to
+use profiler to analyze model performance.
+
+Setup
+-----
+To install ``torch`` and ``torchvision`` use the following command:
+
+::
+
+   pip install torch torchvision
+
+
+"""
+
+
+######################################################################
+# Steps
+# -----
+#
+# 1. Import all necessary libraries
+# 2. Instantiate a simple Resnet model
+# 3. Use profiler to analyze execution time
+# 4. Use profiler to analyze memory consumption
+# 5. Using tracing functionality
+#
+# 1. Import all necessary libraries
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+# In this recipe we will use ``torch``, ``torchvision.models``
+# and ``profiler`` modules.
+#
+
+import torch
+import torchvision.models as models
+import torch.autograd.profiler as profiler
+
+
+######################################################################
+# 2. Instantiate a simple Resnet model
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+# Let's create an instance of a Resnet model and prepare an input
+# for it.
+#
+
+model = models.resnet18()
+inputs = torch.randn(5, 3, 224, 224)
+
+######################################################################
+# 3. Use profiler to analyze execution time
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+# PyTorch profiler is enabled through the context manager and accepts
+# a number of parameters, some of the most useful are:
+#
+# - ``record_shapes`` - whether to record shapes of the operator inputs;
+# - ``profile_memory`` - whether to report amount of memory consumed by
+#   model's Tensors
+# - ``use_cuda`` - whether to measure execution time of CUDA kernels.
+#
+# Let's see how we can use profiler to analyze execution time
+
+with profiler.profile(record_shapes=True) as prof:
+    with profiler.record_function("model_inference"):
+        model(inputs)
+
+######################################################################
+# Note that we can use ``record_function`` context manager to label
+# arbitrary code ranges with user provided names.
```

Comment:
Clarification: Sorry if this is an obvious question, in this example is "model_inference" the label for the code range?

Reply:
will clarify